File: common-parent/ioc/src/main/java/love/forte/common/ioc/annotation/ConfigBeans.java
/*
 * Copyright (c) 2020. ForteScarlet All rights reserved.
 * Project  parent
 * File     ConfigBeans.java
 *
 * You can contact the author through the following channels:
 * github https://github.com/ForteScarlet
 * gitee  https://gitee.com/ForteScarlet
 * email  ForteScarlet@163.com
 * QQ     1149159218
 */

package love.forte.common.ioc.annotation;

import love.forte.common.utils.annotation.AnnotateMapping;
import love.forte.simbot.core.constant.PriorityConstant;

import java.lang.annotation.*;

/**
 *
 * 一般类上可以标记此注解，其默认init为true。
 *
 * @author <a href="https://github.com/ForteScarlet"> ForteScarlet </a>
 */
@Retention(RetentionPolicy.RUNTIME)    //注解会在class字节码文件中存在，在运行时可以通过反射获取到
@Target({ElementType.TYPE, ElementType.METHOD}) //接口、类、枚举、注解、方法
@Documented
@Beans(init = true)
@AnnotateMapping(type = Beans.class)
public @interface ConfigBeans {

    /** 依赖对象的名称，如果没有则以类名取代 */
    String value() default "";

    /** 是否为单例，默认为单例 */
    boolean single() default true;
    /**
     * 是否将类中全部字段标记为Depend，默认为false
     * */
    boolean allDepend() default false;

    /** 当全部标记为@Depend的时候，此参数为所有字段标记的@Depend注解对象，默认为无参注解 */
    Depend depend() default @Depend;


    /** 根据参数类型列表来指定构造函数，默认为无参构造。仅标注在类上的时候有效 */
    Class[] constructor() default {};

    /**
     * 优先级。当在获取某个依赖的时候，假如在通过类型获取的时候存在多个值，会获取优先级更高级别的依赖并摒弃其他依赖。
     * 升序排序。默认为核心最低。
     */
    int priority() default PriorityConstant.CORE_LAST;
}

File: boots/boot-core-annotation/src/main/kotlin/love/forte/simboot/annotation/ConfigBeans.kt
/*
 *  Copyright (c) 2021-2021 ForteScarlet <https://github.com/ForteScarlet>
 *
 *  根据 Apache License 2.0 获得许可；
 *  除非遵守许可，否则您不得使用此文件。
 *  您可以在以下网址获取许可证副本：
 *
 *       https://www.apache.org/licenses/LICENSE-2.0
 *
 *   有关许可证下的权限和限制的具体语言，请参见许可证。
 */

package love.forte.simboot.annotation

import org.springframework.context.annotation.Configuration
import org.springframework.core.annotation.AliasFor

/**
 * 标记一个类为配置bean。
 *
 * 配置bean
 *
 */
@Retention(AnnotationRetention.RUNTIME)
@Target(AnnotationTarget.CLASS)
@MustBeDocumented
@Configuration
public annotation class ConfigBeans(
    @get:AliasFor(annotation = Configuration::class)
    val value: String = ""
)
